File: src/darkroom/preludes/lens.mjs
// lens composition API: call back into this repo's live index/graph from a
// darkroom script without another MCP round-trip.
//
// Every function below execs `lens q <verb> ...` (the read-only CLI family)
// via `execFileSync` with an argv array (never a shell string), parses the
// JSON line `lens q` prints to stdout, and throws on a nonzero exit.
// `LENS_BIN` / `LENS_DIR` are set on this process's environment by the
// darkroom runner that spawned this script.

import { execFileSync } from "node:child_process";

function q(args) {
  const lensBin = process.env.LENS_BIN;
  if (!lensBin) {
    throw new Error(
      "LENS_BIN is not set; lens q calls require the darkroom-injected environment"
    );
  }
  let stdout;
  try {
    stdout = execFileSync(lensBin, ["q", ...args], { encoding: "utf8" });
  } catch (err) {
    const stderr = err.stderr ? String(err.stderr).trim() : "";
    throw new Error(stderr || err.message);
  }
  return JSON.parse(stdout);
}

export function search(query, limit = 20) {
  return q(["search", query, "--limit", String(limit)]);
}

export function symbol(name, kind = null) {
  const args = ["symbol", name];
  if (kind !== null) args.push("--kind", String(kind));
  return q(args);
}

export function callers(name, transitive = false, depth = 2, prodOnly = false) {
  const args = ["callers", name, "--depth", String(depth)];
  if (transitive) args.push("--transitive");
  if (prodOnly) args.push("--prod-only");
  return q(args);
}

export function callees(name, transitive = false, depth = 2, prodOnly = false) {
  const args = ["callees", name, "--depth", String(depth)];
  if (transitive) args.push("--transitive");
  if (prodOnly) args.push("--prod-only");
  return q(args);
}

export function path(frm, to) {
  return q(["path", frm, to]);
}

// `includeBodies`/`withLines` mirror the MCP tool's parameter names, the ones
// models reuse in scripts.
export function skeleton(path, bodies = null, includeBodies = null, withLines = null) {
  const args = ["skeleton", path];
  const b = bodies ?? includeBodies;
  if (b && b.length) args.push("--bodies", b.join(","));
  if (withLines === false) args.push("--no-lines");
  return q(args);
}

export function grep_ast(
  pattern = null,
  query = null,
  lang = null,
  path = null,
  limit = null,
  prodOnly = false,
  language = null
) {
  const args = ["grep-ast"];
  if (pattern !== null) args.push("--pattern", pattern);
  if (query !== null) args.push("--query", query);
  const l = lang ?? language;
  if (l !== null) args.push("--lang", l);
  if (path !== null) args.push("--path", path);
  if (limit !== null) args.push("--limit", String(limit));
  if (prodOnly) args.push("--prod-only");
  return q(args);
}

export function overview(query = null, budget = 8000) {
  const args = ["overview", "--budget", String(budget)];
  if (query !== null) args.push("--query", query);
  return q(args);
}

export function recall(ref, grep = null, offset = null, limit = null) {
  const args = ["recall", ref];
  if (grep !== null) args.push("--grep", grep);
  if (offset !== null) args.push("--offset", String(offset));
  if (limit !== null) args.push("--limit", String(limit));
  return q(args);
}

File: src/darkroom/preludes/lens.py
"""lens composition API: call back into this repo's live index/graph from a
darkroom script without another MCP round-trip.

Every function below execs `lens q <verb> ...` (the read-only CLI family)
via `subprocess.run` with an argv list (never a shell), parses the JSON line
`lens q` prints to stdout, and raises `RuntimeError` on a nonzero exit.
`LENS_BIN` / `LENS_DIR` are set on this process's environment by the
darkroom runner that spawned this script.
"""

import json
import os
import subprocess


def _q(args):
    lens_bin = os.environ.get("LENS_BIN")
    if not lens_bin:
        raise RuntimeError(
            "LENS_BIN is not set; lens q calls require the darkroom-injected environment"
        )
    result = subprocess.run(
        [lens_bin, "q", *args],
        capture_output=True,
        text=True,
    )
    if result.returncode != 0:
        raise RuntimeError(result.stderr.strip() or f"lens q {args[0]} exited {result.returncode}")
    return json.loads(result.stdout)


def search(query, limit=20):
    return _q(["search", query, "--limit", str(limit)])


def symbol(name, kind=None):
    args = ["symbol", name]
    if kind is not None:
        args += ["--kind", str(kind)]
    return _q(args)


def callers(name, transitive=False, depth=2, prod_only=False):
    args = ["callers", name, "--depth", str(depth)]
    if transitive:
        args.append("--transitive")
    if prod_only:
        args.append("--prod-only")
    return _q(args)


def callees(name, transitive=False, depth=2, prod_only=False):
    args = ["callees", name, "--depth", str(depth)]
    if transitive:
        args.append("--transitive")
    if prod_only:
        args.append("--prod-only")
    return _q(args)


def path(frm, to):
    return _q(["path", frm, to])


def skeleton(path, bodies=None, include_bodies=None, with_lines=None, query=None, only=None):
    # `include_bodies`/`with_lines`/`query`/`only` mirror the MCP tool's
    # parameter names, the ones models reuse in scripts (19/28 mined script
    # errors were kwarg mismatches against this shim).
    args = ["skeleton", path]
    bodies = bodies if bodies is not None else include_bodies
    if bodies:
        args += ["--bodies", ",".join(bodies)]
    if with_lines is False:
        args.append("--no-lines")
    if query is not None:
        args += ["--query", query]
    if only is not None:
        args += ["--only", only]
    return _q(args)


def grep_ast(
    pattern=None, query=None, lang=None, path=None, limit=None, prod_only=False, language=None
):
    args = ["grep-ast"]
    if pattern is not None:
        args += ["--pattern", pattern]
    if query is not None:
        args += ["--query", query]
    lang = lang if lang is not None else language
    if lang is not None:
        args += ["--lang", lang]
    if path is not None:
        args += ["--path", path]
    if limit is not None:
        args += ["--limit", str(limit)]
    if prod_only:
        args.append("--prod-only")
    return _q(args)


def overview(query=None, budget=8000):
    args = ["overview", "--budget", str(budget)]
    if query is not None:
        args += ["--query", query]
    return _q(args)


def recall(ref, grep=None, offset=None, limit=None):
    args = ["recall", ref]
    if grep is not None:
        args += ["--grep", grep]
    if offset is not None:
        args += ["--offset", str(offset)]
    if limit is not None:
        args += ["--limit", str(limit)]
    return _q(args)
